sdk/metric/aggregator/minmaxsumcount/mmsc.go

```diff
@@ -104,64 +94,44 @@ func (c *Aggregator) Min() (core.Number, error) {
 func (c *Aggregator) Max() (core.Number, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	if c.checkpoint().count.IsZero(core.Uint64NumberKind) {
+	if c.checkpoint.count.IsZero(core.Uint64NumberKind) {
 		return c.kind.Zero(), aggregator.ErrNoData
 	}
-	return c.checkpoint().max, nil
+	return c.checkpoint.max, nil
 }
 
 // Checkpoint saves the current state and resets the current state to
 // the empty set.
 func (c *Aggregator) Checkpoint(ctx context.Context, desc *metric.Descriptor) {
-	c.lock.SwapActiveState(c.resetCheckpoint)
-}
-
-// checkpoint returns the "cold" state, i.e. state collected prior to the
-// most recent Checkpoint() call
-func (c *Aggregator) checkpoint() *state {
-	return &c.states[c.lock.ColdIdx()]
+	c.lock.Lock()
+	c.checkpoint, c.current = c.current, c.emptyState()
+	c.lock.Unlock()
 }
 
-func (c *Aggregator) resetCheckpoint() {
-	checkpoint := c.checkpoint()
-
-	checkpoint.count.SetUint64(0)
-	checkpoint.sum.SetNumber(c.kind.Zero())
-	checkpoint.min.SetNumber(c.kind.Maximum())
-	checkpoint.max.SetNumber(c.kind.Minimum())
+func (c *Aggregator) emptyState() state {
+	kind := c.kind
+	return state{
+		count: core.NewUint64Number(0),
+		sum:   kind.Zero(),
+		min:   kind.Maximum(),
+		max:   kind.Minimum(),
+	}
 }
 
 // Update adds the recorded measurement to the current data set.
 func (c *Aggregator) Update(_ context.Context, number core.Number, desc *metric.Descriptor) error {
 	kind := desc.NumberKind()
 
-	cIdx := c.lock.Start()
-	defer c.lock.End(cIdx)
-
-	current := &c.states[cIdx]
-	current.count.AddUint64Atomic(1)
-	current.sum.AddNumberAtomic(kind, number)
-
-	for {
-		cmin := current.min.AsNumberAtomic()
-
-		if number.CompareNumber(kind, cmin) >= 0 {
-			break
-		}
-		if current.min.CompareAndSwapNumber(cmin, number) {
-			break
-		}
+	c.lock.Lock()
```

Comment:
Is there a reason to not defer the unlock here to prevent future errors?

Reply:
Not really - other than some possibly misguided concern about performance of defer when the intervening statements are never expected to return prematurely.  I'll change it because I do generally like defer Unlock at the point of locking.